src/json/def_to_json.cr

```diff
@@ -0,0 +1,176 @@
+module JSON
+  # The `def_to_json` macro generates a `to_json(json : JSON::Builder)` method based on provided *mappings*.
+  #
+  # It is a lightweight alternative to `JSON.mapping` if you don't need to declare instance variables and a parser.
+  #
+  # The generated method invoks `to_json(JSON::Builder)` on each of the values returned
+  # by `self`'s properties from *mappings*, or - if a converter is specified - `to_json(value, JSON::Builder)` on the converter.
+  #
+  # ### Example
+  #
+  # ```
+  # require "json"
+  #
+  # record Location, lat : Float64, long : Float64 do
+  #   JSON.def_to_json([lat, long])
+  # end
+  #
+  # record House, street : String, street_number : Int32, location : Location do
+  #   JSON.def_to_json(
+  #     address: true,
+  #     loc: {property: location},
+  #     empty_field: {emit_null: true},
+  #   )
+  #
+  #   def address
+  #     "#{street} #{street_number}"
+  #   end
+  #
+  #   def empty_field
+  #     nil
+  #   end
+  # end
+  #
+  # house = House.new("Crystal Road", 1234, Location.new(12.3, 34.5))
+  # house.to_json # => %({"address":"Crystal Road 1234","loc":{"lat":12.3,"long":34.5},"empty_field":null})
+  # ```
+  #
+  # ### Usage
+  #
+  # `JSON.def_to_json` must receive a series of named arguments, or a named tuple literal, or a hash literal,
+  # whose keys will define JSON properties.
+  #
+  # The value of each key can either be `true` or a hash or named tuple literal with the following options:
+  # * **property**: the property name on the Crystal object (as opposed to the key in the in the JSON document)
+  # * **emit_null**: if `true`, emits a `null` value if the property value is nil (by default nulls are not emitted)
+  # * **converter**: specify an alternate type for generation. The converter must define `to_json(value, JSON::Builder)` as class methods. Examples of converters are `Time::Format` and `Time::EpochConverter` for `Time`.
+  # * **root**: assume the value is inside a JSON object with a given key
+  macro def_to_json(mappings)
+    def to_json(json : ::JSON::Builder)
+      json.object do
+        {% for key, options in mappings %}
+          {% unless options.is_a?(HashLiteral) || options.is_a?(NamedTupleLiteral) %}
+            {% options = {nil: nil} %}
```

Comment:
@RX14 won't work: https://carc.in/#/r/2iyw, @straight-shoota wants the assign a literal, so it can be manipulated from macros